### PR TITLE
Adding Login via Meetup

### PR DESCRIPTION
**Pull request by wbushey**
Added Meetup oAuth provider, DRYed up oAuth path functions, Got rid of Ruby Syntax checker warnings in user model

TODO: Handle not getting a proper email address from Meetup

Addresses #74

### CODE REVIEW

File: app/helpers/application_helper.rb [hunk 1]
```diff
@@ -81,4 +82,11 @@ def provider_link(provider_user)
       provider_user.provider.capitalize
     end
   end
+
+  private
+  def oauth_path(provider, event_id)
```
Comment (pete2786) [inline app/helpers/application_helper.rb]: :+1:

**Comment by wbushey** [comment]
Branch can now get an email address from the user and confirm it by sending a link to that provided address.

Some remaining TODOs:
 - Think of what to do if a user enters an email address that is already associated with a user (anybody have thoughts on that?)
 - Add a little more test coverage
 - Maybe provide a way for the recipient to contest the confirmation.

**Comment by pete2786** [comment]
@wbushey awesome work this weekend! I have been hanging out with my sick kid so it's been great to see these features rolling in. Particularly email confirmation, which is vital to Twitter login as well.

For todos:
#1: I think we can worry about this in the future, I think we are good for now.
#2: Always good, but could be future PRs, have not fully reviewed yet
#3: Good idea as well, I think worth consideration in the future as usage grows.

**Comment by wbushey** [comment]
Thanks @pete2786, and that sounds good. I realized there is one other, I think real todo: the DB migration should set all existing users to have email_confirmed = true.

**Comment by wbushey** [comment]
OK, I've added the migration to set email_confirmed = true for all existing users.

**Comment by pete2786** [comment]
Tested things out locally here, looking pretty good. One thing I'd like to think about is getting rid of the user_id in the route. Couple improvements that we could make.

* Do not allow that action to load if the user already has an email confirmed
* Add a guid/uid to our user table, use that as the identifier for these actions OR POST to this action somehow and do not allow GET (haven't thought this through fully)
* Not needed in this PR, but I realized we need a default image if the provider doesn't give us one
* The sign in modal got a little messy with the new provider, going to have to make some tweaks there

Great work overall, very cool features.

**Comment by wbushey** [comment]
Hi @pete2786, are any of the things you mentioned earlier blockers for this PR? I'd really like to see this Meetup login running on coeventer.com in time for the beginning of Geo:Code 2016 registration, which is going to start next week.

**Comment by pete2786** [comment]
@wbushey The only issue I see as blocking is the second bullet point "Add a guid/uid to our user table, use that as the identifier for these actions OR POST to this action somehow and do not allow GET (haven't thought this through fully)". Right now, someone could 'discover' a user as the only thing needed to get to this view is the user id (browsing directly to /set_email/[:id] for any user. Let me give it a minute of thought tonight, we could merge and the immediately fix this issue too if this is urgent.

**Comment by wbushey** [comment]
I see what you're saying; there's currently no protection on set_email. A couple of things worth noting:

 * `set_email` is following the same convention as `edit`: both are namespaced under /user/{user_id}/ . The big difference is that `edit` checks that the current user is authorized to view/edit the page, while `set_email` does not. I think I ended up here because I changed `current_user` to search within the `email_confirmed` scope, since now we want only users with confirmed email addresses to be able to do the things that any logged in user use to be able to do. In retrospect, changing `current_user` was probably the wrong thing to do; perhaps the Ability model should check for confirmed email instead.
* the user table already has a uid field, and that is currently being set to be the uid provided by oAuth upon creation, both in master and in this PR. I'm pretty sure ProviderUser.uid took on the former purpose of User.uid use to be, but I'm not 100% certain.

**Comment by pete2786** [comment]
By uid I just mean a token or something that is not guessable.

**Comment by pete2786** [comment]
I think we have an easier solution here before going overkill with security. I think we can just render a view that you cannot get to unless you have just tried to sign in with an unconfirmed user. We can get this ready to roll.

**Comment by pete2786** [comment]
Have a potential solution working locally here, I'll push it tomorrow after I test.

**Comment by pete2786** [comment]
I merged this and added my changes to master in this commit, it's deployed, give it a try:

https://github.com/campuscodefest/ccf/commit/267a0337ef4c0af185fd43b7f0567a130db5e442

**Comment by wbushey** [comment]
:thumbsup:

**Pull request merged**